### ISSUE TITLE
White Screen flickering on game launch and exit

### ISSUE THREAD

**Issue by oomek** [mickelson/attract#296]
When I launch a game or exit I get a white screen flashing. Here is the video showing the problem. It happens on any theme.
https://www.youtube.com/watch?v=aA3vEKfg3tI

**Comment by oomek** [comment]
It gets even weirder. When I select a theme and then go back to the previous one I was using the screen flicker is not white but contains a screen of the previous theme. It's like there is some unerrased screen buffer that gets displayed whenever I launch a game.

https://www.youtube.com/watch?v=hu0nacHoQhE

ps. to see it you need to launch an emulator in your desktop resolution and refresh rate.

**Comment by oomek** [comment]
There is one additional flicker the first time I press down after exiting a game.

**Comment by oomek** [comment]
I found out that it happens only when I use Fulscreen. When I set the AM to borderless it does not flicker.

**Comment by mickelson** [comment]
Please post your system info, including OS

**Comment by oomek** [comment]
I5 2500K, radeon r7 260x, 8gb ddr3, ssd, lg 29um67p ultrawide, Windows 10 x64

**Comment by Blakeg612** [comment]
Came here with the same issue. Running borderless seems to "fix" the initial start for a game, but exiting a game leaves the prior screen in place for about 5 seconds with no input response, then a flash to the desktop (including the Attract Mode console window), then a white flash and back to Attract Mode where I was before.

I'm running AM on older hardware (Core 2 Duo, 4 GB RAM) on top of Windows 7 Ultimate. Also using an old MSI TwinFrozr III GeForce 660 card through HDMI.

Any thoughts? Thanks!

**Comment by djrobx** [comment]
Yep I see this too (Windows 10).  It see it mainly when using Kega Fusion, because it leaves a window open for a long time before going into full screen mode. With my other emulators the white flashes are so fast I barely see them.

It kind of seems like Windows wants AM to redraw its UI, but the UI is busy waiting for the emulator process to exit.  Mala displays a "Loading..." screen during Kega's delay.

**Comment by djrobx** [comment]
I tried switching to "borderless".   This greatly improves Kega's load and exit.   However it causes Mame to sometimes hang on exit on this system.   Mame closes its window, so you see AM's wheel, but MAME doesn't fully terminate its process.   As soon as you alt-tab or something, mame exits.   I was able to "catch" this by logging into another RDP session to see that mame64.exe was still actually "running".   I got a crash dump of what Mame was doing, Direct3D seems to get stuck trying to close out. 

win32u!NtUserSystemParametersInfo+0x14
00000000`0a267a60 00007fff`923db23b user32!RealSystemParametersInfoA+0xfc
(Inline Function) --------`-------- uxtheme!ClassicSystemParametersInfoA+0x29 [d:\rs1\shell\themes\uxtheme\sethook.cpp @ 1505]
(Inline Function) --------`-------- uxtheme!_InternalSystemParametersInfo+0x29 [d:\rs1\shell\themes\uxtheme\nctheme.cpp @ 8302]
00000000`0a268020 00007fff`94d8379b uxtheme!ThemeSystemParametersInfoA+0xcb [d:\rs1\shell\themes\uxtheme\sethook.cpp @ 1437]
user32!SystemParametersInfoA+0x9b
d3d9!CEnum::RestoreSystemParameters+0x8d
d3d9!CEnum::DoneExclusiveMode+0x8a
d3d9!handleActivateApp+0x104
d3d9!WindowProc+0x1bb
user32!UserCallWinProcCheckWow+0x274
user32!DispatchClientMessage+0xa7
user32!_fnDWORD+0x33
ntdll!KiUserCallbackDispatcherContinue
win32u!NtUserDestroyWindow+0x14

Both of these problems seem to suggest that AM somehow isn't "cooperating" with the OS while it's waiting for the child process to finish.

**Comment by mickelson** [comment]
Can you try the latest nightly build to see if it fixes the white flicker?: https://build.btolab.com/project/attractmode/

**Comment by oomek** [comment]
Yes, It's almost fixed. The backbuffer blink happens only once on first launch of the game after you start AM and when you alt+tab, go back to AM and then launch a game.

**Comment by djrobx** [comment]
The updated code definitely improves the white flash before game start problem.    Now I only see a short white flicker instead of a persistent white screen with Kega.

**Comment by djrobx** [comment]
Even with the updated code I'm seeing some brief flashes of previous displays (I.e my NES display flickers on briefly when I load or exit a MAME game).  Very odd...

**Comment by oomek** [comment]
Yes, It still flickers unfortunately. White flash on launching groovymame and desktop flash on exiting.

**Comment by oomek** [comment]
Here's the video showing the current status of that flickering for borderless and fullscreen:
https://youtu.be/u98ncL3rldQ

**Comment by mickelson** [comment]
ok that looks like "fullscreen mode" and borderless window.... does the flickering happen on "fill screen" as well?

**Comment by oomek** [comment]
Yes, but only when I exit a game.

**Comment by djrobx** [comment]
I'm using borderless window but similar effect in fullscreen. Both start and exit.

**Comment by djrobx** [comment]
I'm on windows 10 so fill screen not a good option for me.

**Comment by oomek** [comment]
Can you at least ad an option to hide the console window so I can hide the desktop icons and taskbar and set my wallpaper to black? That would be a good workaround.

**Comment by djrobx** [comment]
There is a setting to hide the console window already.   Problem the "white flash" sometimes turns into a flash of a previous screen, and blacking out the desktop doesn't help with this.

**Comment by djrobx** [comment]
Here's a video of what I'm seeing:

https://www.youtube.com/watch?v=RVEQTj2QMY8

You can see my NES display flicker twice.  This is with the latest code, too.

**Comment by djrobx** [comment]
I created a small app that draws a blank, black screen (popup window) then launches AM.   By doing this, I can use Fill Screen (default) in AM, and all the problems go away.   

GameEx takes a similar strategy (HideOS.exe).
